File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        dp = [0]*n
        for i in range(n):
            if i >= k:
                dp[i] = max(0,dp[i-k]) + energy[i]
            else:
                dp[i] = energy[i]
        #print(dp)
        return max(dp[n-k:])

File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    int maximumEnergy(vector<int>& energy, int k) {
        int n = energy.size();
        vector<int> dp(n);
        int ans = INT_MIN;
        for ( int i = 0 ; i < n ; i++)
        {
            if (i>=k)
            {
                dp[i] = max(0,dp[i-k]) + energy[i];
            }
            else{
                dp[i] = energy[i];
            }
            if (i >= n-k) ans = max(ans,dp[i]);
        }
        return ans;
    }
};